common/txmgr/finalizer.go

```diff
@@ -0,0 +1,235 @@
+	f.initSync.Lock()
+	defer f.initSync.Unlock()
+	if f.isStarted {
+		return errors.New("Finalizer is already started")
+	}
+
+	f.stopCh = make(chan struct{})
+	f.wg = sync.WaitGroup{}
+	f.wg.Add(1)
+	go f.runLoop()
+	f.isStarted = true
+	return nil
+}
+
+// Close is a comment to appease the linter
+func (f *Finalizer[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE, HEAD]) Close() error {
+	return f.StopOnce("Finalizer", func() error {
+		return f.closeInternal()
+	})
+}
+
+func (f *Finalizer[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE, HEAD]) closeInternal() error {
+	f.initSync.Lock()
+	defer f.initSync.Unlock()
+	if !f.isStarted {
+		return fmt.Errorf("Finalizer is not started: %w", services.ErrAlreadyStopped)
+	}
+	close(f.stopCh)
+	f.wg.Wait()
+	f.isStarted = false
+	return nil
+}
+
+func (f *Finalizer[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE, HEAD]) Name() string {
+	return f.lggr.Name()
+}
+
+func (f *Finalizer[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE, HEAD]) HealthReport() map[string]error {
+	return map[string]error{f.Name(): f.Healthy()}
+}
+
+func (f *Finalizer[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE, HEAD]) runLoop() {
+	defer f.wg.Done()
+	ctx, cancel := f.stopCh.NewCtx()
+	defer cancel()
+	for {
+		select {
+		case <-f.mb.Notify():
+			for {
+				if ctx.Err() != nil {
+					return
+				}
+				head, exists := f.mb.Retrieve()
+				if !exists {
+					break
+				}
+				if err := f.ProcessHead(ctx, head); err != nil {
+					f.lggr.Errorw("Error processing head", "err", err)
+					f.SvcErrBuffer.Append(err)
+					continue
+				}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (f *Finalizer[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE, HEAD]) ProcessHead(ctx context.Context, head types.Head[BLOCK_HASH]) error {
+	ctx, cancel := context.WithTimeout(ctx, processHeadTimeout)
+	defer cancel()
+	return f.processHead(ctx, head)
+}
+
+// Determines if any confirmed transactions can be marked as finalized by comparing their receipts against the latest finalized block
+func (f *Finalizer[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE, HEAD]) processHead(ctx context.Context, head types.Head[BLOCK_HASH]) error {
+	latestFinalizedHead := head.LatestFinalizedHead()
+	// Cannot determine finality without a finalized head for comparison
+	if latestFinalizedHead == nil || !latestFinalizedHead.IsValid() {
+		return fmt.Errorf("failed to find latest finalized head in chain")
+	}
+	earliestBlockNumInChain := latestFinalizedHead.EarliestHeadInChain().BlockNumber()
+	f.lggr.Debugw("processing latest finalized head", "block num", latestFinalizedHead.BlockNumber(), "block hash", latestFinalizedHead.BlockHash(), "earliest block num in chain", earliestBlockNumInChain)
+
+	// Retrieve all confirmed transactions, loaded with attempts and receipts
+	confirmedTxs, err := f.txStore.FindTransactionsByState(ctx, TxConfirmed, f.chainId)
```

Comment:
Good point. I was trying to use a generic method so that this would be easier to integrate with the in-memory store later on. But that's getting ahead of myself. I've updated to use a more targeted query. I still need to think on batching though. You're right about the large number of tx we'd finalize when this first runs but I also want to avoid slowing this process down for normal use.

Reply:
I've added limits to the batch RPC call to respect the `RPCDefaultBatchSize` but don't think we need to worry about loading all of the confirmed tx into memory. I can discuss more offline in the thread.